Mathlib/Data/Finite/Defs.lean

```diff
@@ -50,11 +50,32 @@ open Function
 
 variable {α β : Sort*}
 
-/-- A type is `Finite` if it is in bijective correspondence to some
-`Fin n`.
-
-While this could be defined as `Nonempty (Fintype α)`, it is defined
-in this way to allow there to be `Finite` instances for propositions.
+/-- A type is `Finite` if it is in bijective correspondence to some `Fin n`.
+
+This is similar to `Fintype`, but `Finite` is a proposition rather than data.
+A particular benefit to this is that `Finite` instances are definitionally equal
+(due to proof irrelevance) rather than being merely propositionally equal.
+Furthermore, `Finite` instances generally avoid needing `Decidable` instances.
+
+Every `Fintype` instance provides a `Finite` instance via `Finite.of_fintype`.
+Conversely, one can noncomputably create a `Fintype` instance from a `Finite` instance
+via `Fintype.ofFinite`. In a proof one might write
+```lean
+  have := Fintype.ofFinite α
+```
+to obtain such an instance.
+
+Do not write noncomputable `Fintype` instances; instead write `Finite` instances
+and use this `Fintype.ofFinite` interface.
+The `Fintype` instances should be relied upon to be computable for evaluation purposes.
+
+Theorems should use `Finite` instead of `Fintype`, unless definitions in the theorem statement
+require `Fintype`.
+Definitions should prefer `Finite` as well, unless it is important that the definitions
+are meant to be computable in the reduction or `#eval` sense.
+
+Implementation note: while this could be defined as `Nonempty (Fintype α)`,
+it is defined in the current way to allow there to be `Finite p` instances for `p : Prop`.
```

Comment:
The last half of this last line is not just an implementation note; the fact that `Finite` works on propositions is a user-visible feature, and very useful for working with indexed finite collections, such as when using `Finsum`.

Reply:
Ok, how about "design note"?

Reply:
I think it's worth having near the top: Finite is a) itself prop not data, and b) works on proof types as well a data types. Both are important API features, and often when you generalize a theorem for a) you should also try to do it for b)

Reply:
Ok, updated